### PR TITLE
Fixed duplicate column names in `merge()` when by.x in names(y)

### PR DESCRIPTION
**Pull request by sritchie73**
When joining two `data.tables` using `merge()` the resulting `data.table` will contain duplicate column names if `by.x != by.y` and `by.x` is also in `names(y)`.

An example:

```{r}
# Create example data.frames
parents <- data.table(name=c("Sarah", "Max", "Qin", "Lex"), 
                      sex=c("F", "M", "F", "M"), 
                      age=c(41, 43, 36, 51))
children <- data.table(parent=c("Sarah", "Max", "Qin"), 
                       name=c("Oliver", "Sebastian", "Kai-lee"),
                       sex=c("M", "M", "F"), 
                       age=c(5,8,7))

# Merge() creates a duplicated "name" column:
merge(parents, children, by.x = "name", by.y = "parent")
```
Output:
```
   name sex.x age.x      name sex.y age.y
1   Max     M    43 Sebastian     M     8
2   Qin     F    36   Kai-lee     F     7
3 Sarah     F    41    Oliver     M     5
```

This behaviour is also present in `base:::merge.data.frame()`, but throws an additional warning:

```
Warning message:
In merge.data.frame(parents, children, by.x = "name", by.y = "parent") :
  column name ‘name’ is duplicated in the result
```

This patch fixes this problem by checking for names shared between `by.x` and `names(y)`, and adding the appropriate `suffixes` to those column names.

### CODE REVIEW

**Comment by sritchie73** [comment]
Also added an equivalent warning message to base for cases when duplicate column names are returned. With the patch above, the only way this can happen is if the user supplies identical values to the suffixes argument (*e.g.* `suffixes=c("", "")`).

File: inst/tests/tests.Rraw [hunk 1]
```diff
@@ -11671,6 +11671,16 @@ unlink(f)
...
+# PR #2631 - Fix duplicated names arising in merge when by.x in names(y)
+# Test should fail in there are any duplicate names after a join
+parents = data.table(name=c("Sarah", "Max"), sex=c("F", "M"), age=c(41, 43))
+children = data.table(parent=c("Sarah", "Max", "Max"), 
+                       name=c("Oliver", "Sebastian", "Michelle"),
+                       sex=c("M", "M", "F"), age=c(5,8,7))
+joined = merge(parents, children, by.x="name", by.y="parent")                       
+test(1877.1, length(names(joined)), length(unique(names(joined)))) 
+test(1877.2, merge(parents, children, by.x="name", by.y="parent", suffixes=c("",""),
+     warning = "column names 'name', 'sex', 'age' are duplicated in the result"),
```
Comment (HughParsonage) [inline inst/tests/tests.Rraw]: Trailing `,` should be removed.

**Comment by sritchie73** [comment]
RE: failed checks: I can't work out how to write the test that checks for the appropriate warning message - would appreciate advice here since there's no documentation for data.table:::test().

I also can't run the tests on my local machine because the package fails to compile (I don't have OpenMP).

**Comment by mattdowle** [comment]
Looks good.  On `test()`, search `tests.Rraw` for `warning=` to see other examples in other tests to follow. You can also pass `output=` and/or `error=` to search for substrings in those messages.
It should compile ok without OpenMP as it should detects if OpenMP is available and switches out.  How are you compiling and what is the message it fails with at what point?  This is my dev cycle: [cc.R]( https://github.com/Rdatatable/data.table/blob/master/cc.R)

**Comment by sritchie73** [comment]
Thanks @mattdowle, I couldn't get the package to build using RStudio package build system - but I probably have something misconfigured in my environment.  I was able to get the tests to work with a bit of trial and error.

**Comment by MichaelChirico** [comment]
I added a little bit to the [Contributing](https://github.com/Rdatatable/data.table/wiki/Contributing) wiki to make it a bit easier to understand how the testing regime works. Thanks for the PR.

**Comment by MarkusBonsch** [comment]
I love it, although it breaks consistency with `merge.data.frame`.  There are quite some other issues with merge.data.table at the moment (#2593). Once these are all fixed (I am working on it), we have to update the documentation to clearly reflect, where we have differences to `merge.data.frame`.

**Comment by MichaelChirico** [comment]
@sritchie73 I assume the Rstudio issue is the same as #2585. I've been using the command line to re-install lately

**Comment by sritchie73** [comment]
Thanks @MarkusBonsch , for what its worth I've also sent a similar patch to the R-devel mailing list for merge.data.frame - but the two people who responded (neither core dev team) were pessimistic about my chances of getting the patch accepted.

@MichaelChirico I get a different error:

```
==> R CMD INSTALL --no-multiarch --with-keep.source data.table

* installing to library ‘/Library/Frameworks/R.framework/Versions/3.4/Resources/library’
* installing *source* package ‘data.table’ ...
** libs
clang -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG   -I/usr/local/include  -fopenmp -fPIC  -Wall -g -O2  -c assign.c -o assign.o
clang: error: unsupported option '-fopenmp'
make: *** [assign.o] Error 1
ERROR: compilation failed for package ‘data.table’
* removing ‘/Library/Frameworks/R.framework/Versions/3.4/Resources/library/data.table’
* restoring previous ‘/Library/Frameworks/R.framework/Versions/3.4/Resources/library/data.table’

Exited with status 1.
```

In the end I managed to get data.table:::test() working almost the same way in my local R session, by copying the file preamble, then fixed the remaining issues that came up in Travis after pushing the changes out to my branch.

**Comment by sritchie73** [comment]
Thanks @jangorecki - I had used `paste` rather than `paste0` in the first instance to match the code style for the existing suffix handling which immediately precedes my patch. Would you prefer I change it to `paste0`, or keep it consistent with the code preceding it?

**Comment by jangorecki** [comment]
Lets wait for R-devel, paste is not that important

**Comment by mattdowle** [comment]
Given Martin Maechler's reply today, it's looking likely to be accepted for R. Wow! Well navigated! I updated manual page accordingly (including link to thread too as Jan suggested) and will merge.
Will invite you to be project member too.

**Comment by sritchie73** [comment]
Thanks!

I am following up with Martin to clarify the functionality of the proposed `no.dups` argument so we can mirror the behaviour of `merge.data.frame`.

**Pull request merged**